Sources/NIOHPACK/StringRing.swift

```diff
@@ -0,0 +1,355 @@
+    }
+    
+    /// The total capacity of this `CircularByteBuffer`.
+    var capacity: Int {
+        return Int(self._storage.capacity)
+    }
+    
+    /// Change the capacity to exactly `to` bytes.
+    ///
+    /// - parameters:
+    ///     - to: The desired minimum capacity.
+    /// - precondition: Int cannot shrink beyond the number of bytes
+    ///                 currently in the buffer.
+    mutating func changeCapacity(to newCapacity: Int) {
+        precondition(newCapacity >= self.readableBytes,
+                     "new capacity \(newCapacity) is too small to contain current buffer contents")
+        if newCapacity == self._storage.capacity {
+            return
+        }
+        
+        // we know this capacity is ok
+        self._reallocateStorageAndRebase(capacity: newCapacity)
+    }
+    
+    /// This vends a pointer to the storage of the `CircularByteBuffer`. It's marked as _very unsafe_ because it might contain
+    /// uninitialised memory and it's undefined behaviour to read it. In most cases you should use `withUnsafeReadableBytes`.
+    ///
+    /// - warning: Do not escape the pointer from the closure for later use.
+    @_inlineable @_versioned
+    func withVeryUnsafeBytes<T>(_ body: (UnsafeRawBufferPointer) throws -> T) rethrows -> T {
+        return try self._storage.withVeryUnsafeBytes(body)
+    }
+    
+    /// Resets the state of this `StringRing` to the state of a freshly allocated one, if possible without
+    /// allocations. This is the cheapest way to recycle a `StringRing` for a new use-case.
+    mutating func clear() {
+        self._storage.clear()
+        self.moveTail(to: 0)
+        self.moveHead(to: 0)
+        self._readableBytes = 0
+    }
+    
+    /// Write `bytes`, a `ContiguousCollection` of `UInt8` into this `StringRing`. Moves the writer index forward by the number of bytes written.
+    /// This method is likely more efficient than the one operating on plain `Collection` as it will use `memcpy` to copy all the bytes in one go.
+    ///
+    /// - parameters:
+    ///     - bytes: A `ContiguousCollection` of `UInt8` to be written.
+    /// - returns: The number of bytes written or `bytes.count`.
+    @discardableResult
+    @_inlineable @_versioned
+    mutating func write<C: ContiguousCollection>(bytes: C) throws -> Int where C.Element == UInt8 {
+        let totalCount = bytes.count
+        guard totalCount <= self.writableBytes else {
+            throw RingBufferError.BufferOverrun(amount: Int(totalCount - self._storage.capacity))
+        }
+        
+        let firstPart = self._storage.capacity - self._ringTail
+        if firstPart >= bytes.count {
+            _storage.set(bytes: bytes, at: self._ringTail)
+        } else {
+            let secondPart = bytes.count - firstPart
+            bytes.withUnsafeBytes { ptr in
+                _storage.set(bytes: UnsafeRawBufferPointer(start: ptr.baseAddress!, count: firstPart), at: self._ringTail)
```

Comment:
can be replaced with this:

```swift
_storage.set(bytes: ptr[..< ptr.startIndex + firstPart], at: self._ringTail)
_storage.set(bytes: ptr[(ptr.startIndex + firstPart) ...], at: 0])
```

which looks nicer

Reply:
Very much agreed.